### PR TITLE
[Executorch] Add MLX recipe

### PR DESCRIPTION
**Pull request by metascroy**
Adds backend="mlx" to the existing Transformers ExecuTorch exporter for Apple Silicon GPU inference, supporting export() and export_for_generation() with static or dynamic shapes.

* Prepares the model on CPU without changing its dtype, disables gradients, and makes input tensors contiguous. Registers cache pytrees before preparation so cropped cache tensors are handled correctly on the first export.
* Uses ExecuTorch’s MLXPartitioner, default MLX transformation passes, and backend-specific edge compilation settings.
* Rejects StaticCache, including static caches wrapped in EncoderDecoderCache, with guidance to use DynamicCache.
* Makes the shared Tensor.reshape patch traceable under strict export while preserving symbolic dimensions and handling non-contiguous inputs. Restricts the no-op detach patch to XNNPACK and CUDA.
* Extends the shared export tests to MLX and adds coverage for backend preparation, configuration, reshape behavior, import isolation, and first-export cache handling. MLX runtime tests explicitly skip when the native MLXBackend is not registered.
* Documents MLX usage, runtime requirements, and known limitations.

Scope Uses the existing generation decomposition and explicit cache inputs/outputs. This PR does not add runtime-owned KV caches, custom attention or MoE implementations, or runtime metadata generation.

Export requires an ExecuTorch installation containing the MLX backend. Execution additionally requires the native MLX delegate and its Metal libraries; operator support depends on the installed backend.

Validation

* Exporter utility tests: 29 passed, 1 skipped, and 65 subtests passed.
* Qwen3 MLX forward and generation export tests: 4 passed, with 14 subtests passed.
* The same four MLX tests correctly skipped when native delegate registration was mocked absent.

### CODE REVIEW

**Comment by metascroy** [comment]
cc @IlyasMoutawwakil here's an MLX recipe.

I've tested llama1b (full attention), gemma3-1b (mixed full/sliding attention), and gemma4-2e (yoco + mixed attention) with fp16/bf16 dtypes and offgraph/in-graph KV cache. In all cases, the models lower to efficient ptes.

What is the quantization story for these recipes? I can follow-up with a PR that supports loading from quantized model-id checkpoints (using torchao format).

File: docs/source/en/exporters.md [hunk 1]
```diff
@@ -165,6 +165,65 @@ outputs = method.execute(list(inputs.values()))
...
+    backend="mlx",
+    cache_mode="in-graph",  # or "off-graph"
+    max_context_len=2048,
+    max_seq_len=512,
+    dtype="bf16",
+    logits_to_keep="last",
+    strict=True,
+    prefer_deferred_runtime_asserts_over_guards=True,
+)
+program = ExecutorchExporter().export(model, {}, config=config)
```
Comment (IlyasMoutawwakil) [inline docs/source/en/exporters.md:190]: where are the inputs ? also why is this using .export and not export_for_generation which is cache-aware ?

File: src/transformers/exporters/exporter_executorch.py [hunk 2]
```diff
@@ -145,8 +146,16 @@ def export(
...
+                from executorch.backends.mlx.passes import get_default_passes
+
+                transform_passes = get_default_passes()
             edge_program_manager: EdgeProgramManager = to_edge_transform_and_lower(
-                exported_program, partitioner=partitioner, compile_config=_get_edge_compile_config()
+                exported_program,
+                partitioner=partitioner,
+                compile_config=_get_edge_compile_config(config.backend),
+                transform_passes=transform_passes,
             )
```
Comment (IlyasMoutawwakil) [inline src/transformers/exporters/exporter_executorch.py:153]: write it as a helper func that takes config.backend, like `compile_config=_get_edge_compile_config(config.backend)`

File: src/transformers/exporters/exporter_executorch.py [hunk 3]
```diff
@@ -259,9 +270,32 @@ def prepare_for_cuda(model: PreTrainedModel, sample_inputs: dict[str, Any]):
...
+        if any(isinstance(cache, StaticCache) for cache in caches):
+            raise ValueError(
+                "StaticCache is not supported by the ExecuTorch MLX backend. "
+                "Use DynamicCache or set cache_implementation='dynamic' in GenerationConfig."
+            )
+
+    from executorch.backends.mlx import MLXPartitioner
+
+    # Traverse cache tensors on the first export, before Dynamo registers their pytrees.
+    register_cache_pytrees_for_model(model)
```
Comment (IlyasMoutawwakil) [inline src/transformers/exporters/exporter_executorch.py]: why ?

File: src/transformers/exporters/exporter_executorch.py [hunk 4]
```diff
@@ -259,9 +270,32 @@ def prepare_for_cuda(model: PreTrainedModel, sample_inputs: dict[str, Any]):
...
+    """Apple Silicon GPU inference via the ExecuTorch MLX backend."""
+    for value in sample_inputs.values():
+        caches = [value]
+        if isinstance(value, EncoderDecoderCache):
+            caches = [value.self_attention_cache, value.cross_attention_cache]
+        if any(isinstance(cache, StaticCache) for cache in caches):
+            raise ValueError(
+                "StaticCache is not supported by the ExecuTorch MLX backend. "
+                "Use DynamicCache or set cache_implementation='dynamic' in GenerationConfig."
+            )
```
Comment (IlyasMoutawwakil) [inline src/transformers/exporters/exporter_executorch.py:286]: any specific reason ?

File: src/transformers/exporters/exporter_executorch.py [hunk 5]
```diff
@@ -635,7 +670,31 @@ def patch(self, *sizes, **kwargs):
...
+        if args or set(kwargs) != {keyword}:
+            raise TypeError(f"Expected positional dimensions or a single {keyword}= argument")
+        args = (kwargs[keyword],)
+    if not args:
+        raise TypeError(f"Missing required {keyword} argument")
+    return args[0] if len(args) == 1 and isinstance(args[0], tuple | list) else args
+
+
+@register_patch("executorch", "torch.Tensor.reshape")
+def _patch_tensor_reshape(_original):
```
Comment (IlyasMoutawwakil) [inline src/transformers/exporters/exporter_executorch.py]: why is torch.Tensor.reshape patch diff than that of torch.reshape ?

File: tests/exporters/test_export.py [hunk 6]
```diff
@@ -730,24 +760,31 @@ def test_onnx_export(self, dynamic):
...
+        model_classes = [
+            model_class
+            for model_class in self.all_model_classes
+            if not any(
+                self._should_skip(model_class, dynamic=dynamic, backend=scope) for scope in ("executorch", backend)
+            )
+        ]
+        if not model_classes:
+            self.skipTest("No model classes support this export configuration")
+        for model_class in model_classes:
```
Comment (IlyasMoutawwakil) [inline tests/exporters/test_export.py]: just call should_skip in the loop like the other backends

File: tests/exporters/test_export.py [hunk 7]
```diff
@@ -892,26 +929,36 @@ def test_onnx_export_generate(self, dynamic, generation_config):
...
+            if not any(
+                self._should_skip(
+                    model_class, generate=True, dynamic=dynamic, backend=scope, generation_config=generation_config
+                )
+                for scope in ("executorch", backend)
+            )
+        ]
+        if not model_classes:
+            self.skipTest("No model classes support this export configuration")
+        for model_class in model_classes:
```
Comment (IlyasMoutawwakil) [inline tests/exporters/test_export.py]: same

File: tests/exporters/test_utils.py
Comment (IlyasMoutawwakil) [inline tests/exporters/test_utils.py:1]: a lot of these tests are unnecessary imo, we test most backend specific things through the end-to-end export, we have many backends and sub-backends so we can't grow backend-specific testing like this, let's keep this file to a minimum of really important tests that actually catch utility regressions from our side.

File: src/transformers/exporters/exporter_executorch.py (hunk 4 above)
Comment (metascroy) [reply src/transformers/exporters/exporter_executorch.py:286]: StaticCache results in user input mutation, which is not supported.

File: src/transformers/exporters/exporter_executorch.py (hunk 3 above)
Comment (metascroy) [reply src/transformers/exporters/exporter_executorch.py]: It was so that _make_contiguous inside prepare backend visits the cache tensors (which requires this registration to happen first), but export on various models succeeds without it so I think it isn't necessary. I'll remove it.

File: src/transformers/exporters/exporter_executorch.py (hunk 5 above)
Comment (metascroy) [reply src/transformers/exporters/exporter_executorch.py]: It shouldn't be, I'll collapse this

File: tests/exporters/test_export.py [hunk 8]
```diff
@@ -576,6 +597,15 @@ class ExportTesterMixin:
...
     """
 
+    def _skip_if_executorch_backend_unavailable(self, backend):
+        if importlib.util.find_spec(f"executorch.backends.{backend}") is None:
+            self.skipTest(f"ExecuTorch backend {backend} is not installed")
+        if backend == "mlx":
+            from executorch.runtime import Runtime
+
+            if "MLXBackend" not in Runtime.get().backend_registry.registered_backend_names:
+                self.skipTest("ExecuTorch runtime does not have the native MLXBackend registered")
```
Comment (IlyasMoutawwakil) [inline tests/exporters/test_export.py]: what about only adding mlx to the _EXECUTORCH_BACKENDS list if mlx is available

File: tests/exporters/test_export.py [hunk 9]
```diff
@@ -61,6 +62,9 @@
 
 
 EXPORT_SKIPS: dict[str, dict[str, str]] = {
+    "mlx.generate": {
+        "Gemma4ForCausalLM": "MLX lowering does not support histc and grouped_mm_fallback in MoE layers.",
```
Comment (IlyasMoutawwakil) [inline tests/exporters/test_export.py]: that's why we model.set_experts_implementation("batched_mm") in prepare xnnpack, please do the same

File: tests/exporters/test_export.py [hunk 10]
```diff
@@ -383,6 +387,23 @@
 )
 
 
+_EXECUTORCH_BACKENDS = ("xnnpack", "mlx")
+EXECUTORCH_EXPORT_PARAMS = parameterized.expand(
+    list(itertools.product(_EXECUTORCH_BACKENDS, _EXPORT_SHAPE_MODES)),
+    name_func=lambda f, _, p: (
+        f"{f.__name__}_{'dynamic' if p.args[1] else 'static'}" + (f"_{p.args[0]}" if p.args[0] != "xnnpack" else "")
```
Comment (IlyasMoutawwakil) [inline tests/exporters/test_export.py]: why not leave the _xnnpack suffix in the test name ?

File: tests/exporters/test_export.py [hunk 11]
```diff
@@ -383,6 +387,23 @@
...
+    ),
+)
+EXECUTORCH_GENERATE_EXPORT_PARAMS = parameterized.expand(
+    list(itertools.product(_EXECUTORCH_BACKENDS, _EXPORT_SHAPE_MODES, _EXPORT_GENERATION_CONFIGS)),
+    name_func=lambda f, _, p: (
+        f"{f.__name__}_{'dynamic' if p.args[1] else 'static'}"
+        + (f"_{p.args[2].cache_implementation}_cache" if p.args[2] is not None else "")
+        + (f"_{p.args[0]}" if p.args[0] != "xnnpack" else "")
+    ),
+)
```
Comment (IlyasMoutawwakil) [inline tests/exporters/test_export.py:613]: can be filtered from mlx x static cache at this point

File: tests/exporters/test_export.py [hunk 12]
```diff
@@ -1081,24 +1115,25 @@ def test_openvino_export(self, dynamic, atol=1e-3, rtol=1e-3):
...
         exporter = ExecutorchExporter()
-        config = ExecutorchConfig(dynamic=dynamic)
+        config = ExecutorchConfig(backend=backend, dynamic=dynamic)
 
         for model_class in self.all_model_classes:
-            if self._should_skip(model_class, dynamic=dynamic, backend="executorch"):
+            if any(
+                self._should_skip(model_class, dynamic=dynamic, backend=scope) for scope in ("executorch", backend)
+            ):
                 continue
```
Comment (IlyasMoutawwakil) [inline tests/exporters/test_export.py:1136]:
```suggestion
            if self._should_skip(model_class, dynamic=dynamic, backend="executorch"):
                continue
            if self._should_skip(model_class, dynamic=dynamic, backend=backend):
                continue
```

File: tests/exporters/test_export.py [hunk 13]
```diff
@@ -1282,26 +1317,28 @@ def test_openvino_export_generate(self, dynamic, generation_config):
...
         for model_class in self.all_generative_model_classes:
-            if self._should_skip(
-                model_class, generate=True, dynamic=dynamic, backend="executorch", generation_config=generation_config
+            if any(
+                self._should_skip(
+                    model_class, generate=True, dynamic=dynamic, backend=scope, generation_config=generation_config
+                )
+                for scope in ("executorch", backend)
             ):
                 continue
```
Comment (IlyasMoutawwakil) [inline tests/exporters/test_export.py]:
```suggestion
            if self._should_skip(
                model_class, generate=True, dynamic=dynamic, backend="executorch", generation_config=generation_config
                ):
            if self._should_skip(
                model_class, generate=True, dynamic=dynamic, backend=backend, generation_config=generation_config
                ):
```

File: tests/exporters/test_export.py [hunk 14]
```diff
@@ -1282,26 +1317,27 @@ def test_openvino_export_generate(self, dynamic, generation_config):
...
         for model_class in self.all_generative_model_classes:
             if self._should_skip(
                 model_class, generate=True, dynamic=dynamic, backend="executorch", generation_config=generation_config
-            ):
-                continue
-
+                ):
+            if self._should_skip(
+                model_class, generate=True, dynamic=dynamic, backend=backend, generation_config=generation_config
+                ):
```
Comment (IlyasMoutawwakil) [inline tests/exporters/test_export.py]:
```suggestion
                ):
                continue
            if self._should_skip(
                model_class, generate=True, dynamic=dynamic, backend=backend, generation_config=generation_config
                ):
                continue
```

File: tests/exporters/test_export.py [hunk 15]
```diff
@@ -1282,26 +1317,29 @@ def test_openvino_export_generate(self, dynamic, generation_config):
...
 
         for model_class in self.all_generative_model_classes:
             if self._should_skip(
                 model_class, generate=True, dynamic=dynamic, backend="executorch", generation_config=generation_config
-            ):
+                ):
+                continue
+            if self._should_skip(
+                model_class, generate=True, dynamic=dynamic, backend=backend, generation_config=generation_config
+                ):
```
Comment (IlyasMoutawwakil) [inline tests/exporters/test_export.py]:
```suggestion
            ):
                continue
            if self._should_skip(
                model_class, generate=True, dynamic=dynamic, backend=backend, generation_config=generation_config
            ):
```

File: docs/source/en/exporters.md [hunk 16]
```diff
@@ -197,6 +197,29 @@ outputs = compiled(ov_inputs)
...
+exporter = ExecutorchExporter()
+config = ExecutorchConfig(backend="mlx", dynamic=True)
+
+components = exporter.export_for_generation(model, inputs, config=config)
+```
+
+For [generation](#generative-models), pass cache outputs from one invocation as inputs to the next.
+Operator support depends on the installed MLX backend. Known lowering limitations include
+static-cache mutation outputs and some captured `None` inputs.
+
```
Comment (IlyasMoutawwakil) [inline docs/source/en/exporters.md:222]: should we not put this inside the executorch option

Comment (metascroy) [reply docs/source/en/exporters.md]: Not put what in the ET option? static cache? Doesn't it work with XNNPACK backend?

Comment (IlyasMoutawwakil) [reply docs/source/en/exporters.md]: i mean the whole mlx example, this one is breaking the symmetry with other backends, we can do options Executorch XNNPACK and Executorch MLX instead

Comment (IlyasMoutawwakil) [reply docs/source/en/exporters.md]: see where we put the backend snippets above in the doc file

Comment (IlyasMoutawwakil) [reply docs/source/en/exporters.md]: also please update the installation options to include the raget registery for mlx if there's one

Comment (metascroy) [reply docs/source/en/exporters.md]: I updated the documentation to make MLX inline inside of the ExecuTorch docs. I only introduced options where the two differed (e.g., installation). Let me know if you want even more separation.

How is this tested in CI? I couldn't find the pin jobs for your model tests, or whether you have mac CI runners.

File: docs/source/en/exporters.md [hunk 17]
```diff
@@ -505,6 +533,10 @@ cache handles where they land internally.
 
 ### Static KV cache
 
+> [!NOTE]
+> The ExecuTorch examples in this section, including zero-copy updates and the decode loop,
+> are **XNNPACK-only**. The MLX exporter rejects `StaticCache`; use `DynamicCache` with MLX.
```
Comment (IlyasMoutawwakil) [inline docs/source/en/exporters.md]: this should be in a single place, it's been repeated across the whole file

Comment (metascroy) [reply docs/source/en/exporters.md]: Removed scattered references and kept only in section on static cache.
